src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,273 @@
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<string> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods
+		) {
+			var groupedMethods = methods
+				.GroupBy( static m => m.Parent )
+				.ToDictionary(
+					static g => g.Key,
+					static g => g.Select( static m => m.Source )
+				);
+
+			return new FileCollector( root, groupedMethods );
+		}
+
+		public string CollectSource() {
+			// File-scoped usings:
+			m_out.Append( m_root.Usings.ToFullString() );
+
+			WriteChildren( m_root );
+
+			if( m_preambles.Count != 0 ) {
+				throw new Bug( "left over preambles" );
+			}
+
+			if( m_methods.Count != 0 ) {
+				throw new Bug( "left over methods" );
+			}
+
+			return m_out.ToString();
+		}
+
+		private bool WriteChildren( SyntaxNode node ) {
+			// Early-out if we have no methods left to place.
+			if( m_methods.Count == 0 ) {
+				return false;
+			}
+
+			var preambleCountBefore = m_preambles.Count;
+
+			bool wroteSomething = false;
+
+			// First recurse into child types
+			foreach( var child in node.ChildNodes() ) {
+				bool wroteSomethingHere = child switch {
+					ClassDeclarationSyntax @class => WriteType( @class ),
+					StructDeclarationSyntax @struct => WriteType( @struct ),
+					InterfaceDeclarationSyntax @interface => WriteType( @interface ),
+
+					// Covers record + record struct:
+					RecordDeclarationSyntax @record => WriteType( @record, @record.ClassOrStructKeyword ),
+
+					// Covers file-scoped and block namespaces
+					BaseNamespaceDeclarationSyntax @namespace => WriteNamespace( @namespace ),
+
+					// Ignore any other child node
+					_ => false
+				};
+
+				if( wroteSomethingHere ) {
+					wroteSomething = true;
+				}
+
+				CheckPremableInvariant( wroteSomething, preambleCountBefore );
+			}
+
+			return wroteSomething;
+		}
+
+		private bool WriteNamespace( BaseNamespaceDeclarationSyntax @namespace ) {
+			SyntaxToken? closeBrace;
+
+			switch( @namespace ) {
+				case FileScopedNamespaceDeclarationSyntax ns:
+					m_preambles.AddLast(
+						ns.NamespaceKeyword.ToFullString()
+						+ ns.Name.ToFullString()
+						+ ns.SemicolonToken.ToFullString()

```

Comment:
Something to note in here is that I'm religiously copy+pasting tokens/syntax from the source file into the destination. This avoids introducing inconsistent whitespace, e.g. the wrong kind of newlines etc.

The only new whitespace tokens I ever use are either an empty token list or a single space character (these are cached in statics above). I do create `partial` tokens and insert them obviously. I could copy the source partial token but by always putting `partial` into the generated code the user gets a diagnostic in their code saying "yo, add a partial here" which is ideal.